File: in-python/bowling.py
class Game(object):

    def __init__(self):
        self.__score = 0
        self.__rolls = []

    def __is_spare(self, roll_index:int):
        end_frame = (roll_index % 2) == 1
        ten_pins = (self.__rolls[roll_index] + 
            self.__rolls[roll_index - 1]) == 10
        return end_frame and ten_pins

    def __spare_bonus(self, roll_index:int):
        return self.__rolls[roll_index + 1]

    def __is_strike(self, roll_index:int):
        return self.__rolls[roll_index] == 10 

    def __strike_bonus(self, roll_index:int):
        return self.__rolls[roll_index + 1] + self.__rolls[roll_index + 2]

    def roll(self, pins:int):
        self.__rolls.append(pins)

    def score(self):
        for roll_index, roll in enumerate(self.__rolls):
            self.__score += roll
            if self.__is_spare(roll_index):
                self.__score += self.__spare_bonus(roll_index)
            if self.__is_strike(roll_index):
                self.__score += self.__strike_bonus(roll_index) 
        return self.__score

File: in-javascript/bowling.js
export default class BowlingGame {
    constructor() {
        this.ballRolls = [];
        this.score = 0;
        this.nFrames = 10;
        this.scoreIndex = 0;
    };


    roll(pins) {
        this.ballRolls.push(pins);
    };

    rollMany(pins , times) {
        for (let i = 0; i < times; i++) {
            this.roll(pins);
        }
    };

    rollSpare(){
        this.rollMany(5, 2);
    };

    scoreFrame() {
        if (this.isStrike()) {
            return this.ballRolls[this.scoreIndex]
        }
        let nextIndex = this.scoreIndex + 1;
        return this.ballRolls[this.scoreIndex] + this.ballRolls[nextIndex];
    };

    isStrike(){
        return this.ballRolls[this.scoreIndex] == 10;
    };

    strikeBonus(){
        let startRoll = this.scoreIndex + 1;
        let endRoll = startRoll + 2;
        let bonuses = this.ballRolls.slice(startRoll, endRoll);
        return bonuses.reduce((a,b) => a + b, 0);
    };

    isSpare(){
        return (!this.isStrike()) && (this.scoreFrame() == 10);
    };

    spareBonus(){
        return this.ballRolls[this.scoreIndex + 2];
    };

    incrementScoreIndex() {
        let incr = this.isStrike() ? 1 : 2;
        this.scoreIndex += incr;
    };

    scoreGame() {
        for (let i = 0; i < this.nFrames; i++){
            this.score += this.scoreFrame()
            if (this.isStrike()){
                this.score += this.strikeBonus();                           
            }
            if (this.isSpare()) {
                this.score += this.spareBonus();
            }
            this.incrementScoreIndex()
        }
        return this.score;
    };
};

